controller/contour/rbac.go

```diff
@@ -0,0 +1,413 @@
+}
+
+// ensureRoleBinding ensures a RoleBinding resource with the provided name exists.
+// The RoleBinding will use svcAct for the subject and role for the role reference.
+func (r *Reconciler) ensureRoleBinding(ctx context.Context, name types.NamespacedName, svcAct *corev1.ServiceAccount, role *rbacv1.Role) error {
+	rb := &rbacv1.RoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: name.Namespace,
+			Name:      name.Name,
+		},
+		Subjects: []rbacv1.Subject{rbacv1.Subject{
+			Kind:      "ServiceAccount",
+			APIGroup:  corev1.GroupName,
+			Name:      svcAct.Name,
+			Namespace: svcAct.Namespace,
+		}},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "Role",
+			Name:     role.Name,
+		},
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Namespace: rb.Namespace, Name: rb.Name}, rb); err != nil {
+		if errors.IsNotFound(err) {
+			if err := r.Client.Create(ctx, rb); err != nil {
+				return fmt.Errorf("failed to create role binding %s: %w", rb.Name, err)
+			}
+			r.Log.Info("created role binding", "name", rb.Name)
+			return nil
+		}
+		return fmt.Errorf("failed to get role binding %s: %w", rb.Name, err)
+	}
+	r.Log.Info("role binding exists; skipped adding", "name", rb.Name)
+	return nil
+}
+
+// ensureClusterRoleBinding ensures a ClusterRoleBinding resource with the provided
+// name exists, using roleName for the role reference and svcAct for the subject.
+func (r *Reconciler) ensureClusterRoleBinding(ctx context.Context, name, roleName string, svcAct types.NamespacedName) error {
+	crb := &rbacv1.ClusterRoleBinding{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Subjects: []rbacv1.Subject{rbacv1.Subject{
+			Kind:      "ServiceAccount",
+			APIGroup:  corev1.GroupName,
+			Name:      svcAct.Name,
+			Namespace: svcAct.Namespace,
+		}},
+		RoleRef: rbacv1.RoleRef{
+			APIGroup: rbacv1.GroupName,
+			Kind:     "ClusterRole",
+			Name:     roleName,
+		},
+	}
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: crb.Name}, crb); err != nil {
+		if errors.IsNotFound(err) {
+			if err := r.Client.Create(ctx, crb); err != nil {
+				return fmt.Errorf("failed to create cluster role binding %s: %w", crb.Name, err)
+			}
+			r.Log.Info("created cluster role binding", "name", crb.Name)
+			return nil
+		}
+		return fmt.Errorf("failed to get cluster role binding %s: %w", crb.Name, err)
+	}
+	r.Log.Info("cluster role binding exists; skipped adding", "name", crb.Name)
+	return nil
+}
+
+// ensureRBACRemoved ensures all the necessary RBAC resources for the provided
+// contour do not exist.
+func (r *Reconciler) ensureRBACRemoved(ctx context.Context, contour *operatorv1alpha1.Contour) error {
+	var errs []error
+	ns := contour.Spec.Namespace.Name
+	objectsToDelete := []runtime.Object{}
+	exist, err := r.contoursExistWithSameSpecNs(ctx, contour)
+	if err != nil {
+		return fmt.Errorf("failed to verify if contours exist in namespace %s: %w",
+			contour.Spec.Namespace.Name, err)
+	}
+	if !exist {
```

Comment:
Maybe I'm not reading right, but if the resources don't exist, we then attempt to delete? Is this just a way to make sure it's all gone if something is partially deleted?

Reply:
> if the resources don't exist, we then attempt to delete?

@stevesloka adding to https://github.com/projectcontour/contour-operator/pull/21#discussion_r494462378, `otherContoursExistInSpecNs()` returns a bool whether other `Contours` exist in the same ns as `contour`. If `false`, it's safe to remove ns-scoped RBAC resources. Similar logic is used later in the func for cluster-scoped RBAC resources.